python/fluid/gan.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2018 Yi Wang <yi.wang.2005@gmail.com> All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import fluid
+
+
+def create_generator_parameters():
+    return {"Gw": fluid.tensor(dim=[2, 20]), "Gb": fluid.tensor(dim=[20])}
+
+
+def create_discriminator_parameters():
+    return {"Dw": fluid.tensor(dim=[20, 2]), "Db": fluid.tensor(dim=[2])}
+
+
+def generate(noise, params):
+    return fluid.fc(noise, params["Gw"], params["Gb"])
+
+
+with fluid.defun("train_discriminator", mb, params):
+    fluid.optimize(
+        fluid.mse(
+            fluid.softmax(fluid.fc(mb[0], params["Dw"], params["Db"])), mb[1]),
+        params)
+
+with fluid.defun("train_generator", noise, g_params, d_params):
+    fluid.optimize(
+        fluid.mse(
+            fluid.softmax(
+                fluid.fc(
+                    generate(noise, g_params), d_params["Dw"], d_params["Db"]),
+                fluid.reshape(True, fluid.sizeof(noise)))), g_params)
+
```

Comment:
I think the functional style API is cool, but a little bit noising.

Maybe for function only have one input and one output, we can use `a.func()` style API?

It likes

```
fluid.fc(generate(noise, g_params), d_params["Dw"], d_params["Db"])).softmax()
```

Reply:
However, it is not important. We might implement a prototype first.

Reply:
I am not sure but I feel that `a.func()` is more confusing than a series of function calls.